File: crates/testrunner/harness/fortran/check.f90
! Vybe test harness — Fortran.
!
! Real Fortran: this file compiles with `gfortran` on its own. Like the COBOL
! harness it documents the shape the emitter produces rather than providing a
! subroutine to splice, because a generic `check` would need a separate
! interface per type (integer, real, logical, character) and the emitted
! comparison is a single line either way.
!
! WHY VALUES, NOT PRINTED TEXT.
! The corpus records what Vybe's logging host produced — bare, unpadded values
! like ["8", "9", "false", "true"]. gfortran's list-directed `print *` pads
! ("           8") and writes logicals as `T`/`F`, so comparing printed text
! would fail under gfortran for formatting reasons alone and the differential
! would be worthless. Comparing the VALUE is something both runtimes agree on,
! because it is semantics rather than formatting.
!
! The emitted check, replacing `print *, dst(1)` whose expected line is "8":
!
!     if ((dst(1)) /= 8) then
!         print *, "FAIL: want [8] got [", dst(1), "]"
!         stop 1
!     end if
!
! chosen by the shape of the expected text:
!
!     8        integer   ->  (x) /= 8
!     3.5      real      ->  abs((x) - 3.5) > 1.0e-6      (never compare reals with /=)
!     true     logical   ->  (x) .neqv. .true.
!     anything else      ->  trim(x) /= "..."
!
! FAILURE USES `stop <code>`, which now carries a real status: the walker used
! to lower `stop` to a bare `Return` and the profile mapped it to `noop`, so
! every `stop 1` exited 0 while gfortran gave 1. Measured after the fix:
! `stop 1` -> 1 and `error stop 2` -> 2 in both runtimes.
program vybe_check_shape
    integer :: x
    x = 8
    if ((x) /= 8) then
        print *, "FAIL: want [8] got [", x, "]"
        stop 1
    end if
    print *, "shape ok"
end program vybe_check_shape

File: crates/testrunner/harness/c/check.c
/* Vybe test harness — C.
 *
 * Real C: this compiles with `cc` on its own. Like the COBOL and Fortran
 * harnesses it documents the shape the emitter produces rather than providing
 * a function to splice, because the check has to be inline — see below.
 *
 * WHY THE CHECK IS INLINE AND PER PRINT.
 * Every way of ACCUMULATING output fails under Vybe while working under cc
 * (all measured):
 *
 *   static char g[64] written from a function  ->  undefined is not callable
 *                                                  (__libc_char_to_str)
 *   snprintf(buf + n, ...)  — pointer arithmetic -> js-string.compare: first
 *                                                  arg not a string
 *   strcat(dst, src)                           ->  js-number.toF64: not a number
 *   vsnprintf + va_list                        ->  undefined is not callable
 *
 * So there is no buffer to collect into, and each `printf` is compared where
 * it stands. What DOES work in both: a LOCAL `char t[N]`, `snprintf` writing
 * at offset 0, `strcmp` against a literal, and `printf("%s", t)`.
 *
 * WHY assert(0) AND NOT exit/return.
 * `assert` is the ONLY failure signal that reaches a non-zero status in both:
 *
 *   assert(0)            cc 134   vybex 1    <- the one that works
 *   exit(1) in a block   cc 1     vybex 0
 *   return 3 from main   cc 3     vybex 0
 *   abort()              cc 134   vybex 0
 *
 * `exit_call_is_return = true` (languages/c/src/profile:14) compiles `exit` to
 * a RETURN, and main's return value is not mapped to the process status. Both
 * are real C bugs; until they are fixed `assert` is the only option, and it is
 * enough because the verdict only needs non-zero.
 *
 * The FAIL line is printed BEFORE asserting: an uncaught failure renders as
 * `RuntimeError: [object]` under Vybe, so the values would otherwise be lost.
 */
#include <stdio.h>
#include <string.h>
#include <assert.h>

int main(void) {
    int a[] = {5, 3, 8};

    /* One emitted check, replacing `printf("%d %d %d\n", a[0], a[1], a[2]);`
     * whose expected line is "5 3 8". The braces keep `__t` scoped so repeated
     * checks in one function do not collide. */
    {
        char __t[512];
        snprintf(__t, sizeof(__t), "%d %d %d\n", a[0], a[1], a[2]);
        if (strcmp(__t, "5 3 8\n") != 0) {
            printf("FAIL: want [5 3 8] got [%s]\n", __t);
            assert(0);
        }
    }

    printf("harness ok\n");
    return 0;
}
